lib/Crypto/crypto_c.cpp

```diff
@@ -0,0 +1,1523 @@
+    QByteArray keyData = Sailfish::Crypto::Key::serialise(
+                cppKey, Sailfish::Crypto::Key::LosslessSerialisationMode);
+    const char *keyDataPtr = keyData.data();
+
+    /* Now marshal it for dbus */
+    dbus_message_iter_open_container(
+                in_args,
+                DBUS_TYPE_STRUCT,
+                NULL,
+                &key_struct);
+
+    dbus_message_iter_open_container(
+                &key_struct,
+                DBUS_TYPE_ARRAY,
+                "y",
+                &data_array);
+
+    dbus_message_iter_append_fixed_array(
+                &data_array,
+                DBUS_TYPE_BYTE,
+                &keyDataPtr,
+                keyData.size());
+
+    dbus_message_iter_close_container(
+                &key_struct,
+                &data_array);
+
+    dbus_message_iter_close_container(
+                in_args,
+                &key_struct);
+}
+
+int Sailfish_Crypto__readResultStruct(DBusMessageIter *args,
+                                      struct Sailfish_Crypto_Result **result)
+{
+    if (dbus_message_iter_get_arg_type(args) != DBUS_TYPE_STRUCT) {
+        fprintf(stderr, "Argument is not result struct!\n");
+        if (dbus_message_iter_get_arg_type(args) == DBUS_TYPE_STRING) {
+            char *resultStr;
+            dbus_message_iter_get_basic(args, &resultStr);
+            fprintf(stderr, "Got result str: %s\n", resultStr);
+        }
+        return 0;
+    } else {
+        DBusMessageIter res_struct;
+        int code;
+        int errorCode;
+        int storageErrorCode;
+        char *errorMessage;
+
+        dbus_message_iter_recurse(args, &res_struct);
+
+        if (dbus_message_iter_get_arg_type(&res_struct) != DBUS_TYPE_INT32) {
+            fprintf(stderr, "First result struct arg not an integer!\n");
+            return 0;
+        }
+
+        dbus_message_iter_get_basic(&res_struct, &code);
+
+        dbus_message_iter_next(&res_struct);
+
+        if (dbus_message_iter_get_arg_type(&res_struct) != DBUS_TYPE_INT32) {
+            fprintf(stderr, "Second result struct arg not an integer!\n");
+            return 0;
+        }
+
+        dbus_message_iter_get_basic(&res_struct, &errorCode);
+
+        dbus_message_iter_next(&res_struct);
+
+        if (dbus_message_iter_get_arg_type(&res_struct) != DBUS_TYPE_INT32) {
+            fprintf(stderr, "Third result struct arg not an integer!\n");
+            return 0;
+        }
+
+        dbus_message_iter_get_basic(&res_struct, &storageErrorCode);
+
+        dbus_message_iter_next(&res_struct);
+
+        if (dbus_message_iter_get_arg_type(&res_struct) != DBUS_TYPE_STRING) {
+            fprintf(stderr, "Fourth result struct arg not a string!\n");
+            return 0;
+        }
+
+        dbus_message_iter_get_basic(&res_struct, &errorMessage);
+
+        *result = Sailfish_Crypto_Result_new(
+                    (Sailfish_Crypto_Result_Code)code,
+                    errorCode, storageErrorCode, errorMessage);
```

Comment:
Ok, forget this comment, checking https://dbus.freedesktop.org/doc/api/html/group__DBusMessage.html#ga41c23a05e552d0574d0444d4693d18ab there is no memory leak.

To avoid thinking this when reading the code, would the compiler allow to declare errorMessage as a const char* instead of a char* ?

Reply:
Done.